### PR TITLE
iOS-180 Fix emtpy search result if acquisition of fetched feed entries not supported

### PR DESCRIPTION
**Pull request by ErnestFan**
**What's this do?**
This fixes an empty search result being shown if the acquisition of all the entries are not supported.
We fetch the next page in such case.

**Why are we doing this? (w/ JIRA link if applicable)**
[ios-180](https://jira.nypl.org/browse/IOS-180?filter=-1)

**How should this be tested? / Do these changes have associated tests?**
Go to NYPL library
Search `How to do everything with`
Multiple search results should be shown

**Dependencies for merging? Releasing to production?**
N/A

**Does this include changes that require a new SimplyE/Open eBooks build for QA?**
N/A

**Has the application documentation been updated for these changes?**
N/A

**Did someone actually run this code to verify it works?**
@ErnestFan

### CODE REVIEW

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m [hunk 1]
```diff
@@ -84,6 +91,7 @@ - (instancetype)initWithOPDSFeed:(NYPLOPDSFeed *const)feed
     }
     [self.books addObject:book];
   }
+  self.booksFromLastFetchNotSupported = (self.books.count == 0) && (feed.entries.count > 0);
```
Comment (ettore) [inline Simplified/Catalog/NYPLCatalogUngroupedFeed.m:94]: if I understand well this means that we had some books in the feed but none of them is supported by SimplyE, correct? If so, do you think `noSupportedBooksInLastFetch` is more clear? I was struggling to understand the meaning of this property until I saw the code. 

I was also wondering if it can be a computed property instead -- I'm not sure that's ok though.

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m [hunk 2]
```diff
@@ -176,11 +184,30 @@ - (void)prepareForBookIndex:(NSUInteger)bookIndex
...
+  [self fetchNextPageWithCompletionHandler:nil];
+}
+
+// This method might trigger a recursive loop, if and only if
+// the fetched feed contains more than one entry AND none of the entries are supported AND nextURL is not nil
+// [NYPLCatalogUngroupedFeed withURL:handler] <-> [NYPLCatalogUngroupedFeed fetchNextPageWithCompletionHandler:]
+// The handler should always be called if it is not nil, in order to exit the recursive loop
+- (void)fetchNextPageWithCompletionHandler:(nullable void (^)(NYPLCatalogUngroupedFeed *category))handler
+{
+  if(self.currentlyFetchingNextURL) {
```
Comment (ettore) [inline Simplified/Catalog/NYPLCatalogUngroupedFeed.m:200]: I'm not sure we want to call the handler every time we call this method and we are currently fetching... that would result in a disparity of `handler` call compared to the actual fetch.

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m [hunk 3]
```diff
@@ -195,14 +222,23 @@ - (void)prepareForBookIndex:(NSUInteger)bookIndex
...
+         }
          return;
        }
        
        [self.books addObjectsFromArray:ungroupedFeed.books];
        self.nextURL = ungroupedFeed.nextURL;
        self.currentlyFetchingNextURL = NO;
        
-       [self prepareForBookIndex:self.greatestPreparationIndex];
+       if(handler) {
```
Comment (ettore) [inline Simplified/Catalog/NYPLCatalogUngroupedFeed.m:235]: I could be misunderstanding, but shouldn't the handler be called only when we are exiting/breaking the recursive loop?

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m [hunk 4]
```diff
@@ -176,11 +184,30 @@ - (void)prepareForBookIndex:(NSUInteger)bookIndex
...
-  if(self.currentlyFetchingNextURL) return;
+  if(self.books.count - bookIndex > preloadThreshold) {
+    return;
+  }
   
-  if(!self.nextURL) return;
+  [self fetchNextPageWithCompletionHandler:nil];
+}
+
+// This method might trigger a recursive loop, if and only if
```
Comment (ettore) [inline Simplified/Catalog/NYPLCatalogUngroupedFeed.m:194]: I would rephrase this a bit because as it is it looks like this recursive loop is not intended, but it is. So it's a little bit confusing. How about 
```
Continues to recursively fetch results from `nextURL` if all the books from last fetch are not supported.
@param handler Called at the end of the recursion.
```
?

(let me know if I'm off, I'm not 100% sure what I wrote is correct)

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m (hunk 1 above)
Comment (ErnestFan) [reply Simplified/Catalog/NYPLCatalogUngroupedFeed.m:94]: Right, I need to have the naming convention tips open all the time.

I don't think it can be a computed property as we use the entries of `NYPLOPDSFeed` to determine the value, and we are not storing that in `NYPLCatalogUngroupedFeed`

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m (hunk 3 above)
Comment (ErnestFan) [reply Simplified/Catalog/NYPLCatalogUngroupedFeed.m:235]: From my understanding, we are exiting here because we have found books that are supported and need to return the `NYPLCatalogUngroupedFeed` object to the view controller, in order to update the UI and set the delegates. The view controller will call `prepareForBookIndex:` if more books are needed.

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m (hunk 1 above)
Comment (ettore) [reply Simplified/Catalog/NYPLCatalogUngroupedFeed.m:94]: oh I see, you're right... sounds good.

File: Simplified/Catalog/NYPLCatalogUngroupedFeed.m (hunk 3 above)
Comment (ettore) [reply Simplified/Catalog/NYPLCatalogUngroupedFeed.m:235]: ah, so we are supposed to exit but then it's missing a `return` statement after the `handler` call maybe?

**Review by ettore — APPROVED** [review]
this looks ok for a temporary fix, considering all the limitations discussed in the ticket, whose implications however reach outside the scope of IOS-180. 

So while I am approving this PR (it does fix the bug), I would prefer to wait to merge this until we get the authorization to work on IOS-239 as a quick follow-up. Essentially I don't know if it's a good idea to live with this code for a long time.

**Comment by MalcolmMcFly** [comment]
@ErnestFan what are we going to do with this PR?

**Comment by ErnestFan** [comment]
> @ErnestFan what are we going to do with this PR?

@MalcolmMcFly We probably will close this once we rebuild this module as described in the TAD ([iOS-243](https://jira.nypl.org/browse/IOS-243)). I am leaving this open in case we need to release this temporary fix.

**Comment by ettore** [comment]
@ErnestFan  can we close this one now?

**Comment by ErnestFan** [comment]
@ettore Sure
Note: This can be closed since this part of code has been refactored in this [PR](https://github.com/NYPL-Simplified/Simplified-iOS/pull/1523)
